src/components/context/TodosContext.tsx

```diff
@@ -0,0 +1,210 @@
+
+  useEffect(() => {
+    const fetchTodos = async () => {
+      try {
+        const fetchedTodos = await todosServices.getTodos();
+
+        setTodos(fetchedTodos);
+      } catch (error) {
+        setErrorMessage(TodoError.UnableToLoad);
+      }
+    };
+
+    fetchTodos();
+  }, [setErrorMessage]);
+
+  const addTodo = async ({ title, completed, userId }: Todo) => {
+    setErrorMessage(null);
+
+    setTempTodo({
+      id: 0,
+      title,
+      userId,
+      completed: false,
+    });
+
+    try {
+      const newTodo = await todosServices.postTodos({
+        title,
+        completed,
+        userId,
+      });
+
+      setTodos(currTodos => [...currTodos, newTodo]);
+    } catch (error) {
+      setErrorMessage(TodoError.UnableToAdd);
+      setTempTodo(null);
+      throw error;
+    } finally {
+      setTempTodo(null);
+    }
+  };
+
+  const removeTodo = async (todoId: number) => {
+    setIsLoading(true);
+
+    try {
+      await todosServices.deleteTodos(todoId);
+      setTodos(currTodos => currTodos.filter(todo => todo.id !== todoId));
+    } catch (error) {
+      setErrorMessage(TodoError.UnableToDelete);
+      throw error;
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  const updateTodo = async (updatedTodo: Todo) => {
+    try {
+      const todoEdited = await todosServices.updateTodo(updatedTodo);
+
+      setTodos(currTodos => {
+        const newTodos: Todo[] = [...currTodos];
+        const index = newTodos.findIndex(todo => todo.id === updatedTodo.id);
+
+        newTodos.splice(index, 1, todoEdited);
+
+        return newTodos;
+      });
+    } catch (error) {
+      setErrorMessage(TodoError.UnableUpdate);
+      throw error;
+    }
+  };
+
+  const handleCheck = (updatedTodo: Todo) => {
+    setTodos(prevTodos =>
+      prevTodos.map(todo =>
+        todo.id === updatedTodo.id
+          ? { ...todo, completed: !todo.completed }
+          : todo,
+      ),
+    );
+  };
+
+  const toggleAll = () => {
+    setTodos(prevTodos => {
+      const allCompleted = prevTodos.every(todo => todo.completed);
+
+      return prevTodos.map(todo => ({
+        ...todo,
+        completed: !allCompleted,
+      }));
+    });
+  };
+
+  const handleClearCompleted = async () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+    const failedDeletions: Todo[] = [];
+
+    for (const todo of completedTodos) {
+      try {
+        await todosServices.deleteTodos(todo.id);
+        setIsLoading(false);
+      } catch (error) {
+        setTodos(todos);
+        failedDeletions.push(todo);
+        setErrorMessage(TodoError.UnableToDelete);
+      }
+    }
+
+    if (failedDeletions.length === 0) {
+      setTodos(todos.filter(todo => !todo.completed));
+    } else {
+      setTodos(
+        todos.filter(
+          todo =>
+            !completedTodos.includes(todo) || failedDeletions.includes(todo),
+        ),
+      );
+      setErrorMessage(TodoError.UnableToDelete);
+    }
+  };
```

Comment:
You have a bit of strange and complicated logic, and some things you repeat such as setting an error message.
The flow could look like:
1. Create a bunch of requests.
2. Await for a response with `Promise.all` or `Promise.allSettled`.
3. Handler a response/error/remove loading states only ones